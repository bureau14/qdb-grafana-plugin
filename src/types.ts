import { DataQuery, DataSourceJsonData } from '@grafana/data';

export interface MyQuery extends DataQuery {
  queryText?: string;
  constant: number;
}

export const defaultQuery: Partial<MyQuery> = {
  constant: 6.5,
};

/**
 * Value that is used in the backend, but never sent over HTTP to the frontend
 */
export interface MySecureJsonData {
  apiKey?: string;
}

/**
 * These are options configured for each DataSource instance
 */
export interface MyDataSourceOptions extends DataSourceJsonData {
  uri?: string;
}

/**
 * Value that is used in the backend, but never sent over HTTP to the frontend
 */
export interface SecureJsonData {
  username?: string;
  user_private_key?: string;
  cluster_public_key?: string;
}
