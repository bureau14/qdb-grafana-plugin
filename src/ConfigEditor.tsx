import React, { ChangeEvent, PureComponent } from 'react';
import { LegacyForms } from '@grafana/ui';
import { DataSourcePluginOptionsEditorProps } from '@grafana/data';
import { MyDataSourceOptions, SecureJsonData } from './types';

const { SecretFormField, FormField } = LegacyForms;

interface Props extends DataSourcePluginOptionsEditorProps<MyDataSourceOptions> {}

interface State {}

export class ConfigEditor extends PureComponent<Props, State> {
  onHostChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    const jsonData = {
      ...options.jsonData,
      host: event.target.value,
    };
    onOptionsChange({ ...options, jsonData });
  };
  onUsernameChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    onOptionsChange({
      ...options,
      secureJsonData: {
        ...options.secureJsonData,
        username: event.target.value,
      },
    });
  };
  // Secure field (only sent to the backend)
  onUserPrivateKeyChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { onOptionsChange, options } = this.props;
    onOptionsChange({
      ...options,
      secureJsonData: {
        ...options.secureJsonData,
        user_private_key: event.target.value,
      },
    });
  };

  onResetUsername = () => {
    const { onOptionsChange, options } = this.props;
    onOptionsChange({
      ...options,
      secureJsonFields: {
        ...options.secureJsonFields,
        username: false,
      },
      secureJsonData: {
        ...options.secureJsonData,
        username: '',
      },
    });
  };
  onResetUserPrivateKey = () => {
    const { onOptionsChange, options } = this.props;
    onOptionsChange({
      ...options,
      secureJsonFields: {
        ...options.secureJsonFields,
        user_private_key: false,
      },
      secureJsonData: {
        ...options.secureJsonData,
        user_private_key: '',
      },
    });
  };

  render() {
    const { options } = this.props;
    const { jsonData, secureJsonFields } = options;
    const secureJsonData = (options.secureJsonData || {}) as SecureJsonData;

    return (
      <div className="gf-form-group">
        <div className="gf-form">
          <FormField
            label="Host"
            labelWidth={6}
            inputWidth={20}
            onChange={this.onHostChange}
            value={jsonData.host || ''}
            placeholder="A host ip address"
          />
        </div>
        <div className="gf-form">
          <SecretFormField
            isConfigured={(secureJsonFields && secureJsonFields.username) as boolean}
            label="Username"
            labelWidth={6}
            inputWidth={20}
            onChange={this.onUsernameChange}
            value={secureJsonData.username || ''}
            onReset={this.onResetUsername}
            placeholder="User"
          />
        </div>

        <div className="gf-form-inline">
          <div className="gf-form">
            <SecretFormField
              isConfigured={(secureJsonFields && secureJsonFields.user_private_key) as boolean}
              value={secureJsonData.user_private_key || ''}
              label="User Key"
              placeholder="User Key"
              labelWidth={6}
              inputWidth={20}
              onReset={this.onResetUserPrivateKey}
              onChange={this.onUserPrivateKeyChange}
            />
          </div>
        </div>
      </div>
    );
  }
}
