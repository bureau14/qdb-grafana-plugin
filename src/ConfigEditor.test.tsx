// ConfigEditor.test.js
import React from 'react';
import { create } from 'react-test-renderer';
import { ConfigEditor } from './ConfigEditor';
import { DataSourcePluginOptionsEditorProps } from '@grafana/data';
import { QdbDataSourceOptions, QdbSecureJsonData } from './types';
import { createDatasourceSettings } from './mocks';

test('ConfigEditor', () => {
  let jsonData: QdbDataSourceOptions = {
    host: 'https://127.0.0.1:443',
  };
  let secureJsonData: QdbSecureJsonData = {
    user: 'test-user',
    secret: 'SIqZ+vg4eZGeuhzCyT90OtwweivdtOieodaaTV/Xtd6k=',
  };

  let opts = createDatasourceSettings(jsonData);
  let props: DataSourcePluginOptionsEditorProps<QdbDataSourceOptions> = {
    options: {
      ...opts,
      secureJsonData: secureJsonData,
    },
    onOptionsChange: () => {
      return;
    },
  };

  const component = create(<ConfigEditor options={props.options} onOptionsChange={props.onOptionsChange} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
