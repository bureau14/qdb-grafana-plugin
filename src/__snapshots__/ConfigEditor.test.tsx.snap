// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigEditor 1`] = `
<div
  className="gf-form-group"
>
  <div
    className="gf-form"
  >
    <div
      className="form-field"
    >
      <label
        className="gf-form-label width-6"
      >
        Host
      </label>
      <input
        className="gf-form-input width-20"
        onChange={[Function]}
        placeholder="Quasardb REST API URL"
        type="text"
        value="https://127.0.0.1:443"
      />
    </div>
  </div>
  <div
    className="gf-form"
  >
    <div
      className="form-field"
    >
      <label
        className="gf-form-label width-6"
      >
        Username
      </label>
      <input
        className="gf-form-input width-20"
        onChange={[Function]}
        placeholder="User"
        type="password"
        value="test-user"
      />
    </div>
  </div>
  <div
    className="gf-form-inline"
  >
    <div
      className="gf-form"
    >
      <div
        className="form-field"
      >
        <label
          className="gf-form-label width-6"
        >
          User Key
        </label>
        <input
          className="gf-form-input width-20"
          onChange={[Function]}
          placeholder="User Key"
          type="password"
          value="SIqZ+vg4eZGeuhzCyT90OtwweivdtOieodaaTV/Xtd6k="
        />
      </div>
    </div>
  </div>
</div>
`;
